// dthmysql/index.js
var mysql = require('mysql');
var pool = mysql.createPool({
    connectionLimit: 50,
    host: '192.168.30.204',
    user: 'root',
    password: 'naxiewunai',
    database: 'postgresqltest',
    multipleStatements: true,//是否支持多行语句执行
    supportBigNumbers: true,//数据库支持bigint或decimal类型列时，需要设此option为true
});
function query(sql, callback) {
    pool.getConnection(function (err, connection) {
        connection.query({
            sql, function(field, next) {
                if (field.type == 'VAR_STRING' || field.type == 'BLOB') {
                }
                return next();
            }
        }, function (err, rows) {
            if (err) {
                connection.rollback(function () {
                    var str = {
                        code: 0,
                        message: err.message
                    };
                    callback(null, str);
                    connection.release();
                    return;
                });
            } else {
                var str = "";
                if (rows !== undefined) {
                    if (rows.length > 0) {
                        str = {
                            code: 0,
                            message: "success",
                            data: rows
                        };
                    } else {
                        str = {
                            code: 1,
                            message: "no result"
                        };
                    }

                } else {
                    str = {
                        code: 0,
                        message: "success",
                        data: rows
                    };
                }
                callback(null, str);
                connection.release();
                return;
            }
        }
        );
    });
}
function connect(sql, sql_Params, type, callback) {
    pool.getConnection(function (err, connection) {
        connection.query(sql, sql_Params, function (err, rows) {
            if (err) {
                var str = {//查询失败报错信息
                    code: 0,
                    message: err
                };
                callback(null, str);
                connection.release();
                return;
            }
            else {
                var str = {};
                if (type == "insert") {
                    str = {
                        code: 1,
                        message: {
                            affectedrows: rows.affectedRows,//受影响的行数
                            insertid: rows.insertId//新增的主键ID
                        }
                    };
                } else {
                    str = {
                        code: 1,
                        message: {
                            affectedrows: rows.affectedRows,//受影响的行数
                        }
                    };
                }
                callback(null, str);
                connection.release(); //关闭数据库连接，结束        
                return;
            }
        });
    })
};
exports.dthmysql = {
    query: function (sql) {
        return function (callback) {
            query(sql, callback);
        };
    },
    insert: function (sql, sql_Params) {
        return function (callback) {
            connect(sql, sql_Params, "insert", callback);
        };
    },
    update: function (sql, sql_Params) {
        return function (callback) {
            connect(sql, sql_Params, "update", callback);
        };
    },
    remove: function (sql, sql_Params) {
        return function (callback) {
            connect(sql, sql_Params, "remove", callback);
        };
    }
}
